### PR TITLE
[Inductor] FlexAttention supports partial masking

### PR DESCRIPTION
**Pull request by yanboliang**
Tested on causal using [this script](https://gist.github.com/yanboliang/7c34a82df611d4ea8869cb9e041bfbfc): we got ~20% speedup for both forward and backward compared with no partial mask support.
```
(pt) [ybliang@devgpu002.ash8 ~/local/debug]$ python benchmark.py 
fwd speedup: 0.7345418152465407
bwd speedup: 0.6989407329455135
(pt) [ybliang@devgpu002.ash8 ~/local/debug]$ python benchmark.py --partial-mask
fwd speedup: 0.8731704554826333
bwd speedup: 0.8312823672034853
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: test/inductor/test_flex_attention.py [hunk 1]
```diff
@@ -108,6 +109,11 @@ def _trig2(score, b, h, m, n):
     return z
 
 
+# --------- Useful mask functions for testing ---------
+def _causal_mask(batch, head, token_q, token_kv):
```
Comment (Chillee) [inline test/inductor/test_flex_attention.py:113]: I wonder if we should just make this have the same function signature as `score_mod`. It's a little bit awkward, but arguably having a single function signature makes it simpler.

File: torch/nn/attention/_flex_attention.py [hunk 2]
```diff
@@ -259,24 +303,57 @@ def _create_mask(
...
-    score_mod: _score_mod_signature,
+    fn: Callable,
     B: int,
     H: int,
     M: int,
     N: int,
     device: str = "cuda",
     KV_BLOCK_SIZE: int = _DEFAULT_SPARSE_BLOCK_SIZE,
     Q_BLOCK_SIZE: int = _DEFAULT_SPARSE_BLOCK_SIZE,
+    is_score_mod: bool = True,
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:351]: ehhhhhhhh

File: torch/nn/attention/_flex_attention.py [hunk 3]
```diff
@@ -307,20 +387,34 @@ def _create_empty_block_mask(query, key, value) -> _BlockMask:
...
+) -> torch.Tensor:
+    return token_q >= 0
+
+
 def _flex_attention(
     query: torch.Tensor,
     key: torch.Tensor,
     value: torch.Tensor,
     score_mod: _score_mod_signature = _identity,
+    mask_fn: _mask_signature = _no_mask,
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:417]: I think `mask_fn` should be part of the `block_mask`

File: torch/_inductor/kernel/flex_attention.py [hunk 4]
```diff
@@ -240,13 +200,114 @@ def build_subgraph_buffer(
...
+    # FULL_KV_IDX and FULL_KV_NUM_BLKS are always contiguous.
     sparse_hz_offset = sparse_idx_z * SPARSE_H + sparse_idx_h
     sparse_kv_num_blks_offset = sparse_hz_offset * SPARSE_Q_BLOCK_CNT + start_m // SPARSE_Q_MULTIPLE
     sparse_kv_idx_offset = sparse_hz_offset * SPARSE_Q_BLOCK_CNT * SPARSE_KV_BLOCK_CNT + (start_m // SPARSE_Q_MULTIPLE) * SPARSE_KV_BLOCK_CNT  # noqa: B950
-    kv_indices = SPARSE_KV_IDX + sparse_kv_idx_offset
+    kv_indices = FULL_KV_IDX + sparse_kv_idx_offset
     kv_start = tl.load(kv_indices) * SPARSE_KV_BLOCK_SIZE # first kv block we're loading
-    sparse_kv_num_blocks = tl.load(SPARSE_KV_NUM_BLKS + sparse_kv_num_blks_offset)
+    sparse_kv_num_blocks = tl.load(FULL_KV_NUM_BLKS + sparse_kv_num_blks_offset)
+
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_attention.py:216]: Can we refactor this code to share the code paths for the two branches?

File: test/inductor/test_flex_attention.py (hunk 1 above)
Comment (yanboliang) [reply test/inductor/test_flex_attention.py:113]: Yea, I think if we make ```score_mod``` and ```mask``` have the same signature, it would be simpler.

Comment (yanboliang) [reply test/inductor/test_flex_attention.py:113]: Discussed offline: ```score_mod``` and ```mask_fn``` should be different signatures. For ```create_block_mask``` function, advanced users would provide ```mask_fn``` for better performance, and fall back to non partial mask routine if users don't provide ```mask_fn```.

**Pull request closed without merging**